### PR TITLE
Service CRUD, Fixing bugs on main.py, Refactored

### PR DESCRIPTION
**Pull request by qangdev**
1. Added Admin template (Bootstrap)
2. Services listing, adding, editing
3. Fix bugs in main
4. I added a file named "database.py" it's for centralize the database call. From now on you can import it like from app.database import db and it will work like normal. I make sure that the db.json will only be created under gear directory
5. Refactored models: services, shops, vehicles. I wrap all about functions (insert, delete, all) into a repo class. So basically a Repo class, it's only job is for access database to manipulate data. It won't handle business data

### CODE REVIEW

File: app/database.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import sys
+import tinydb
+
+from pathlib import Path
+
+# DB initializtion
+DBJSON_PATH = Path(sys.path[1]) / Path("db.json")
+print(DBJSON_PATH)
```
Comment (mohanadkaleia) [inline app/database.py]: this seems to be used for debugging, can you remove it if it is not needed anymore

File: app/forms.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from flask_wtf import FlaskForm
+from wtforms import StringField, FloatField, FileField, TextField, MultipleFileField, TextAreaField
```
Comment (mohanadkaleia) [inline app/forms.py:2]: I think flake8 prevents multi imports, can you install the pre-commit hooks using 'pre-commit install',

File: app/forms.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from wtforms.validators import ValidationError
+from wtforms.validators import DataRequired, Length
+
+
+class FormNewServices(FlaskForm):
+    name = TextField('name')
+    description = TextAreaField('description')
+    price = FloatField('price')
```
Comment (mohanadkaleia) [inline app/forms.py:10]: the `price` should be a dict object, not one value, yesterday I shared this doc that illustrates more info about services.. you can double check it from here https://docs.google.com/document/d/1se7TXX3YcZYdZ8TX9MG3ILQFgQIxwv6GWzLdW5Zs1tE/edit?usp=sharing

File: app/models/services.py [hunk 4]
```diff
@@ -12,67 +15,123 @@ class ErrInvalidParameters(Exception):
...
-        raise ErrInvalidParameters("service name must be provided")
-
-    service = tinydb.Query()
-    result = services.search(service.name == name)
-
-    if not result:
-        raise ErrNotFound(f"no service found for the provided name: {name}")
-
-    return result[0]
+class ServiceRepo:
```
Comment (mohanadkaleia) [inline app/models/services.py:18]: hmmm.. interesting, why the name of this class is `ServiceRepo`? I actually prefer `Service`, is there a reason for Repo?

File: app/models/services.py [hunk 5]
```diff
@@ -12,67 +15,123 @@ class ErrInvalidParameters(Exception):
...
+
+    def __init__(self, db: tinydb.TinyDB):
+        self.db = db
+        self.services = db.table('services')
+
+    def insert(self, id="",
+               shop_id="",
+               name="",
+               slug="",
```
Comment (mohanadkaleia) [inline app/models/services.py:27]: I think no need to pass slug as a parameter, we can just generate it on the go inside the insert method

File: app/models/services.py [hunk 6]
```diff
@@ -12,67 +15,123 @@ class ErrInvalidParameters(Exception):
...
+
+    def update(self, id="",
+               name="",
+               slug="",
+               description="",
+               price="",
+               images=None):
+        query = tinydb.Query()
+        doc = {"name": name,
+               "slug": slug,
```
Comment (mohanadkaleia) [inline app/models/services.py:50]: same note here, the slug should be auto generated from the name

File: app/models/services.py [hunk 7]
```diff
@@ -12,67 +15,123 @@ class ErrInvalidParameters(Exception):
...
-        raise ErrNotFound(f"no service found for the provided name: {name}")
-
-    return result[0]
+class ServiceRepo:
+
+    def __init__(self, db: tinydb.TinyDB):
+        self.db = db
+        self.services = db.table('services')
+
+    def insert(self, id="",
```
Comment (mohanadkaleia) [inline app/models/services.py:24]: the id should be auto generated with in the insert, please use the util.random_id func

File: app/static/admin/css/dashboard.rtl.css [hunk 8]
```diff
@@ -0,0 +1,96 @@
+body {
```
Comment (mohanadkaleia) [inline app/static/admin/css/dashboard.rtl.css:1]: interesting, why the `rtl` css is added here?

File: app/static/admin/css/bootstrap.min.css [hunk 9]
```diff
@@ -0,0 +1,7 @@
+@charset "UTF-8";/*!
```
Comment (mohanadkaleia) [inline app/static/admin/css/bootstrap.min.css:1]: I prefer to load bootstrap css file using CDN (I highly discourage having it on disk)

File: app/static/admin/js/bootstrap.bundle.min.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
+/*!
```
Comment (mohanadkaleia) [inline app/static/admin/js/bootstrap.bundle.min.js:1]: same note here, please use CDN to load bootstrap js files

**Comment by mohanadkaleia** [comment]
Thank you, I left few comments on your pr, for the future, can we divide such a pr into smaller prs? for example refactoring models can be considered one pr, creating `admin` can be considered another pr.. the smaller the easier to iterate and fix in case of issues.. 

Thank you

File: app/database.py (hunk 1 above)
Comment (mohanadkaleia) [reply app/database.py]: also why are we putting the database in the sys.path?

**Comment by mohanadkaleia** [comment]
Refactoring the models into classes is not clear to me, I see for each model you created two classes `Repo` and `Service` class, 
1. Why we have two classes? why not just one and handle the 
2. Why the Repo class is used, can't we instead have it to be just functions as before?
3. My method in engineering is to make things as simple as possible, I think we can have it more simpler as it was before 😎but I'm happy to go your way if you can convince me 😝
4. I prefer to have the db to be on the same root dir as it was before, not clear to me why it was moved to the sys directory
5. I liked you created a specific file to initialize the db.. thank you! 
6. oh.. do you use virtualenv? I was surprised that your changes did not work on my local.. I think maybe different python version -- I use pyhton 3.9.1

File: app/models/services.py [hunk 11]
```diff
@@ -12,67 +15,123 @@ class ErrInvalidParameters(Exception):
...
+        result = self.services.search(service.name == name)
+
+        if not result:
+            raise ErrNotFound(f"no service found for the provided name: {name}")
+
+        return result[0]
+
+
+class Service:
+    UPLOADED_DIR_PATH: str = str(Path.cwd() / Path("app/static/uploaded"))
```
Comment (mohanadkaleia) [inline app/models/services.py:92]: For the future, this should be placed in the config.py file (but don't worry about it from now)

File: app/models/services.py [hunk 12]
```diff
@@ -12,67 +15,123 @@ class ErrInvalidParameters(Exception):
...
+        directory = Path(Service.UPLOADED_DIR_PATH) / Path(self.id)
+        directory.mkdir(parents=True, exist_ok=True)
+        return directory
+
+    @property
+    def prefix_images(self):
+        return [Path(self.upload_dir)/Path(image) for image in self.images]
+
+    @classmethod
+    def populate(cls, form: FormNewServices):
```
Comment (mohanadkaleia) [inline app/models/services.py:122]: The model layer should not be aware of how the forms in the front-end are created, instead we should just pass parameters from there.

File: app/models/services.py [hunk 13]
```diff
@@ -12,67 +15,123 @@ class ErrInvalidParameters(Exception):
...
+        return cls(shop_id="FAKE ID:FIX ME",
+                   name=form.name.data,
+                   slug=slugify(form.name.data),
+                   description=form.description.data,
+                   price=form.price.data)
+
+    @classmethod
+    def populate_edit(cls, service, form: FormNewServices):
```
Comment (mohanadkaleia) [inline app/models/services.py:130]: why we used the name populate, this more fit generate a batch of records. Instead can we just use `edit`, `create`... etc?

File: app/database.py (hunk 1 above)
Comment (qangdev) [reply app/database.py]: Yup, I'll delete the print statement
Actually it'll be updated in the next commit

File: app/forms.py (hunk 2 above)
Comment (qangdev) [reply app/forms.py:2]: I'll do it, thanks

File: app/models/services.py (hunk 4 above)
Comment (qangdev) [reply app/models/services.py:18]: There is also a `Service` class in the same file. So me idea is `Service` class is for handling business logic only like on the other hand a Repo is only know about saving, retrieving data from the database. We can combine those classes into one but the Service is coupled to the database layer.
What do you think? I'm ok if you prefer one class only :)

File: app/models/services.py (hunk 5 above)
Comment (qangdev) [reply app/models/services.py:27]: I'll update it , thanks

File: app/models/services.py (hunk 7 above)
Comment (qangdev) [reply app/models/services.py:24]: actually it's still using util.random_id func. So before calling insert function. A Service instance is created the instance's id is generated by util.random_id func in the constructor method.

File: app/static/admin/css/dashboard.rtl.css (hunk 8 above)
Comment (qangdev) [reply app/static/admin/css/dashboard.rtl.css:1]: I'll remove that file

File: app/static/admin/css/bootstrap.min.css (hunk 9 above)
Comment (qangdev) [reply app/static/admin/css/bootstrap.min.css:1]: I'll update it. Thanks

File: app/static/admin/js/bootstrap.bundle.min.js (hunk 10 above)
Comment (qangdev) [reply app/static/admin/js/bootstrap.bundle.min.js:1]: I'll update it. Thanks

File: app/models/services.py (hunk 12 above)
Comment (qangdev) [reply app/models/services.py:122]: I'll update it. Thanks

File: app/models/services.py (hunk 13 above)
Comment (qangdev) [reply app/models/services.py:130]: yup, thank you. Franky it was hard to have to good name for it

**Pull request closed without merging**